### PR TITLE
[Merged by Bors] - feat: port CategoryTheory.Functor.FullyFaithful

### PR DESCRIPTION
**Pull request by rosborn**
mathlib SHA: 70d50ecfd4900dd6d328da39ab7ebd516abe4025

Porting notes:
1. ~~`aesop_cat` was unable to close goals that `tidy` was able to. I left an example proof of what `tidy` emits (with some cleanup).~~
2. `unfold_projs` is not ported. Is it an easy tactic to port over or is there a better "lean4" way? (ed: not needed)
3. `trans` doesn't seem to work with `HEq` (see https://github.com/leanprover-community/mathlib4/issues/1119)
4. Some documentation is needed for `Full` and `Faithful`. Please feel free to fill it in if you wish.

### CODE REVIEW

**Comment by mpenciak** [comment]
I filled in all the errors, and the file typechecks now.

There's a lot of TODOs and mysteries still around why certain things worked and others stopped working...

Needless to say there's a lot of cleanup left to do before we merge, but at least it all works now :)

File: Mathlib/CategoryTheory/Functor/FullyFaithful.lean
Comment (jcommelin) [inline Mathlib/CategoryTheory/Functor/FullyFaithful.lean:5]:
```suggestion

! This file was ported from Lean 3 source module category_theory.functor.fully_faithful
! leanprover-community/mathlib commit 70d50ecfd4900dd6d328da39ab7ebd516abe4025
! Please do not edit these lines, except to modify the commit id
! if you have ported upstream changes.
-/
```

Comment (jcommelin) [reply Mathlib/CategoryTheory/Functor/FullyFaithful.lean:5]: Since very recently, mathport adds these headers automatically. We'll have to add them manually to older PRs.

File: Mathlib/CategoryTheory/Functor/FullyFaithful.lean
Comment (adamtopaz) [inline Mathlib/CategoryTheory/Functor/FullyFaithful.lean:298]: This asymmetry in the names feels strange to me... I'll bring this up on zulip.

**Review by adamtopaz — APPROVED** [review]
LGTM! I'll wait for another maintainer to review since I made some commits to this branch.

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**
